File: LEI 3ano/2 Semestre/PL/TPC1/code.py
import re

def processar_tpc_com_detalhes(nome_ficheiro):
    try:
        with open(nome_ficheiro, 'r', encoding='utf-8') as f:
            # Tokenização por caracteres de espaçamento
            tokens = f.read().split()
    except FileNotFoundError:
        print(f"Erro: O ficheiro '{nome_ficheiro}' não foi encontrado.")
        return

    # Listas para guardar as matches
    matches_min_num = []
    matches_mais_mais = []

    # Regex para o segundo critério: +alfa-numérico+
    pattern_mais = re.compile(r'^\+[a-zA-Z0-9]+\+$')

    for t in tokens:
        # 1. Começa com minúscula e termina com número
        if len(t) >= 2 and t[0].islower() and t[-1].isdigit():
            matches_min_num.append(t)
        
        # 2. Começa e acaba em '+' com alfa-numéricos no meio
        if pattern_mais.match(t):
            matches_mais_mais.append(t)

    # Impressão dos Resultados
    print(f"{'='*50}")
    print(f"RESULTADOS PARA: {nome_ficheiro}")
    print(f"{'='*50}\n")

    print(f"1. Tokens que começam com minúscula e terminam com número ({len(matches_min_num)}):")
    if matches_min_num:
        print(f"   Matches: {matches_min_num}")
    else:
        print("   Nenhuma match encontrada.")
    
    print("\n" + "-"*50 + "\n")

    print(f"2. Tokens entre '+' com conteúdo alfa-numérico ({len(matches_mais_mais)}):")
    if matches_mais_mais:
        print(f"   Matches: {matches_mais_mais}")
    else:
        print("   Nenhuma match encontrada.")
    print(f"\n{'='*50}")

# Execução
processar_tpc_com_detalhes('tpc1.txt')

File: LEI 3ano/2 Semestre/PL/TPC1/code.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>

int match_regex(const char *pattern, const char *text) {
    regex_t regex;
    int result;
    
    if (regcomp(&regex, pattern, REG_EXTENDED | REG_NOSUB) != 0) {
        return 0;
    }
    
    result = regexec(&regex, text, 0, NULL, 0);
    regfree(&regex);
    
    return result == 0;
}

int main() {
    FILE *file = fopen("tpc1.txt", "r");
    if (file == NULL) {
        printf("Erro ao abrir o ficheiro\n");
        return 1;
    }
    
    char token[1024];
    int count1 = 0, count2 = 0;
    
    // Pergunta 1: Tokens que começam com letra minúscula e terminam com dígito
    // Regex: ^[a-z].*[0-9]$
    const char *regex1 = "^[a-z].*[0-9]$";
    
    // Pergunta 2: Tokens que começam e acabam com '+' e têm apenas alfanuméricos entre eles
    // Regex: ^\+[a-zA-Z0-9]+\+$
    const char *regex2 = "^\\+[a-zA-Z0-9]+\\+$";
    
    while (fscanf(file, "%s", token) != EOF) {
        if (match_regex(regex1, token)) count1++;
        if (match_regex(regex2, token)) count2++;
    }
    
    fclose(file);
    
    printf("Pergunta 1 - Tokens que começam com letra minúscula e terminam com dígito: %d\n", count1);
    printf("Pergunta 2 - Tokens que começam e acabam com '+' e têm apenas alfanuméricos entre eles: %d\n", count2);
    
    return 0;
}